File: backend/app/api/v1/scheduled_events.py
"""Scheduled event endpoints — deferred event delivery."""

import uuid

from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import ApiKeyDep, SessionDep, api_key_filter_id
from app.models.enums import ScheduledEventStatus
from app.models.scheduled_event import ScheduledEvent
from app.schemas.common import PaginatedResponse
from app.schemas.scheduled_event import ScheduledEventCreate, ScheduledEventResponse
from app.services import scheduled_event_service

router = APIRouter(prefix="/scheduled-events", tags=["scheduled-events"])


def _to_response(event: ScheduledEvent) -> ScheduledEventResponse:
    payload = event.payload or {}
    return ScheduledEventResponse(
        id=event.id,
        api_key_id=event.api_key_id,
        event_type=str(payload.get("event_type") or ""),
        scheduled_for=event.scheduled_for,
        priority=event.priority,
        status=event.status,
        event_id=event.event_id,
        created_at=event.created_at,
        updated_at=event.updated_at,
    )


@router.post("", response_model=ScheduledEventResponse, status_code=status.HTTP_201_CREATED)
async def create_scheduled_event(
    body: ScheduledEventCreate,
    *,
    db: SessionDep,
    api_key: ApiKeyDep,
) -> ScheduledEventResponse:
    """Schedule an event for deferred delivery at a future timestamp."""
    event = await scheduled_event_service.create_scheduled_event(db, body, api_key.id)
    return _to_response(event)


@router.get("", response_model=PaginatedResponse[ScheduledEventResponse])
async def list_scheduled_events(
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=20, ge=1, le=100),
    status_filter: ScheduledEventStatus | None = Query(default=None, alias="status"),
    *,
    db: SessionDep,
    api_key: ApiKeyDep,
) -> PaginatedResponse[ScheduledEventResponse]:
    api_key_filter = api_key_filter_id(api_key)
    events, total = await scheduled_event_service.list_scheduled_events(
        db, api_key_filter, status=status_filter, page=page, per_page=per_page
    )
    return PaginatedResponse.create([_to_response(e) for e in events], total, page, per_page)


@router.delete("/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
async def cancel_scheduled_event(
    event_id: uuid.UUID,
    *,
    db: SessionDep,
    api_key: ApiKeyDep,
) -> None:
    """Cancel a pending scheduled event."""
    api_key_filter = api_key_filter_id(api_key)
    event = await scheduled_event_service.cancel_scheduled_event(db, event_id, api_key_filter)
    if event is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Scheduled event not found",
        )
    if event.status != ScheduledEventStatus.CANCELLED:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot cancel event with status '{event.status}'",
        )

File: packages/sdk/src/scheduled-events.ts
import { mapRecipient } from "./events";
import { type ApiPage, HttpClient, mapPage, query } from "./http";
import { CreateScheduledEventInputSchema } from "./schemas";
import type {
  CreateScheduledEventInput,
  EventPriority,
  Page,
  RequestOptions,
  ScheduledEvent,
  ScheduledEventListOptions,
  ScheduledEventStatus,
} from "./types";

type ApiScheduledEvent = {
  id: string;
  api_key_id: string;
  event_type: string;
  scheduled_for: string;
  priority: EventPriority;
  status: ScheduledEventStatus;
  event_id: string | null;
  created_at: string;
  updated_at: string;
};

function mapScheduledEvent(value: ApiScheduledEvent): ScheduledEvent {
  return {
    id: value.id,
    apiKeyId: value.api_key_id,
    eventType: value.event_type,
    scheduledFor: value.scheduled_for,
    priority: value.priority,
    status: value.status,
    eventId: value.event_id,
    createdAt: value.created_at,
    updatedAt: value.updated_at,
  };
}

/** Creates, lists, and cancels deferred events. */
export class ScheduledEventsResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * Schedules an event for future delivery.
   * @param input Event content and future delivery time.
   * @param options Optional cancellation signal.
   * @returns The created scheduled event.
   * @throws A `BeacoError` when validation or the request fails.
   */
  async create(
    input: CreateScheduledEventInput,
    options: RequestOptions = {},
  ): Promise<ScheduledEvent> {
    input = CreateScheduledEventInputSchema.parse(input);
    const value = await this.http.request<ApiScheduledEvent>("/scheduled-events", {
      method: "POST",
      body: {
        event_type: input.eventType,
        recipients: input.recipients.map(mapRecipient),
        scheduled_for:
          input.scheduledFor instanceof Date
            ? input.scheduledFor.toISOString()
            : input.scheduledFor,
        priority: input.priority,
        template_id: input.templateId,
        payload: input.payload,
        metadata: input.metadata,
      },
      signal: options.signal,
    });
    return mapScheduledEvent(value);
  }

  /**
   * Lists scheduled events visible to the configured project API key.
   * @param options Filters, pagination, and optional cancellation signal.
   * @returns One page of scheduled events.
   * @throws A `BeacoError` when the request fails.
   */
  async list(options: ScheduledEventListOptions = {}): Promise<Page<ScheduledEvent>> {
    const page = await this.http.request<ApiPage<ApiScheduledEvent>>(
      `/scheduled-events${query(options)}`,
      { signal: options.signal },
    );
    return mapPage(page, mapScheduledEvent);
  }

  /**
   * Cancels a pending scheduled event.
   * @param id Scheduled-event identifier.
   * @param options Optional cancellation signal.
   * @returns Nothing after successful cancellation.
   * @throws A `BeacoError` when the event cannot be cancelled or the request fails.
   */
  async cancel(id: string, options: RequestOptions = {}): Promise<void> {
    await this.http.request<void>(`/scheduled-events/${id}`, {
      method: "DELETE",
      signal: options.signal,
    });
  }
}
